### PR TITLE
Current form takes `#_` commented forms with cursor in all positions

### LINKED ISSUE #3115: When we are in a commented form in any cursor position  both the marker and the ignored form must be considered as Current Form
https://github.com/BetterThanTomorrow/calva/pull/3111#issuecomment-3979704874

Update the current form for all these:

```clojure
|#_:a
#|_:a
#_|:a
#_:|a
#_:a|
```

To be:

```clojure
|#_:a|
```

### PR DESCRIPTION
**Pull request by jramosg**
## What has changed?

- Current form includes #_ at all cursor positions
- nsFromCursorDoc no longer treats `#_` discarded ns forms as valid 
    - All three code paths in `nsRangeFromCursorDoc` now check if the   
  form range starts with an ignore token and skip it. This prevents `#_` discarded forms
   from being treated as valid ns declarations.

Fixes #3115

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Made sure there is an issue registered with a clear problem statement that this PR addresses, (created the issue if it was not present).
    - [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [x] Figured if the change might have some side effects and tested those as well.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

### CODE REVIEW

**Comment by jramosg** [comment]
this test fails
```ts

    // TODO: Figure what to do with ignored forms
    //       For now, this is what they do (nothing)
    it('Finds ns in top level ignored form', function () {
      expect(nsFormUtil.nsFromCursorDoc(docFromTextNotation('#_ (ns a-b.c-d)|'))).toStrictEqual([
        'a-b.c-d',
        '(ns a-b.c-d)',
      ]);
    });
```

it's receiving `["a-b.c-d", "#_ (ns a-b.c-d)"]`
I think ns for `#_ (ns a-b.c-d)` should be `null` what do you think @PEZ ?

**Comment by PEZ** [comment]
Yeah, `null` is fine. Is that what you get without fixing anything?

**Comment by jramosg** [comment]
I had that "error" when I did the change about treating #_ form as current form at all cursor positions.
Now I changed the `nsRangeFromCursorDoc` so it does not tre`at #_(ns x) ` as valid ns

**Comment by PEZ** [comment]
Sweet. I've tested it and It feels pretty natural. There's some inconsistency now between selecting the current form and moving forward and backward a form at a time, as well as with dragging pairs. But it is not like we did not have inconsistencies before. I think we are in a much better place now. Thanks! 🙏

File: src/extension-test/unit/cursor-doc/token-cursor-test.ts [hunk 1]
```diff
@@ -773,6 +773,36 @@ describe('Token Cursor', () => {
...
+      const expected = textAndSelection(b)[1];
+      for (const notation of ['|#_(foo bar)', '#|_(foo bar)', '#_|(foo bar)']) {
+        const a = docFromTextNotation(notation);
+        const cursor: LispTokenCursor = a.getTokenCursor(a.selections[0].anchor);
+        expect(cursor.rangeForCurrentForm(a.selections[0].anchor)).toEqual(expected);
+      }
+    });
+    it('selects only the keyword when cursor is in whitespace after #_:a in #_:a :b', () => {
+      const a = docFromTextNotation('#_:a| :b');
+      const b = docFromTextNotation('#_|:a| :b');
```
Comment (jramosg) [inline src/extension-test/unit/cursor-doc/token-cursor-test.ts]: what do you think about this @PEZ ?
if I change it to take `|#_:a|` this test should be changed too, and I think we don't want that
```ts
   // https://github.com/BetterThanTomorrow/calva/issues/2655
      it('Does not include ignore marker', () => {
        const a = docFromTextNotation('aaa (comment #_ [bbb ccc|]  ddd)');
        const b = docFromTextNotation('aaa (comment #_ |[bbb ccc]|  ddd)');
        const cursor: LispTokenCursor = a.getTokenCursor(a.selections[0].active);
        expect(cursor.rangeForDefun(a.selections[0].active)).toEqual(textAndSelection(b)[1]);
      });
```

But maybe this is inconsistent with 
`#_:a|` -> `|#_:a|` ? And it should be `#_|:a|` ?

Comment (PEZ) [reply src/extension-test/unit/cursor-doc/token-cursor-test.ts]: I think we should include the ignore marker in both cases. The whitespace shouldn't make a difference.

Comment (jramosg) [reply src/extension-test/unit/cursor-doc/token-cursor-test.ts]: yes, whitespace does not make difference, I've whitespaces tested
```ts
   it('selects ignore form including #_ when cursor is at end of #_     :a (spaces between)', () => {
      const a = docFromTextNotation('#_     :a|');
      const b = docFromTextNotation('|#_     :a|');
      const cursor: LispTokenCursor = a.getTokenCursor(a.selections[0].anchor);
      expect(cursor.rangeForCurrentForm(a.selections[0].anchor)).toEqual(textAndSelection(b)[1]);
    });
```

It it when we have a form after.  I thought it was intentional after reading  https://github.com/BetterThanTomorrow/calva/issues/2655
So I will change it 👌

Comment (PEZ) [reply src/extension-test/unit/cursor-doc/token-cursor-test.ts]: Good catch! It's important that we keep supporting #2655. This is what I referred to when I said I couldn't quite grasp the implications of this type of change. Let me think a bit... brb!

Comment (jramosg) [reply src/extension-test/unit/cursor-doc/token-cursor-test.ts]: aah okey, so I will try to support that in `rangeForDefun` while selections works as `#_:a| :b -> |#_:a| :b`

Comment (PEZ) [reply src/extension-test/unit/cursor-doc/token-cursor-test.ts]: I think we will need some sort of special case. Similar, in spirit, to how we have a special idea of top-level when inside Rich comments.

Let's see where you arrive with your idea of handling it in `rangeForDefun`. Sounds like the right approach to me.

File: src/cursor-doc/token-cursor.ts [hunk 2]
```diff
@@ -815,7 +832,10 @@ export class LispTokenCursor extends TokenCursor {
           return lastCandidateRange;
         }
       } else {
-        lastCandidateRange = cursor.rangeForCurrentForm(cursor.offsetStart);
+        const end = cursor.offsetStart;
```
Comment (jramosg) [inline src/cursor-doc/token-cursor.ts:835]: @PEZ this is the `rangeForDefun` case, it cn be managed  with `backwardSexp` and skipping comments
I also added a test to check when there is not whitespace between the  #_ and the form

**Comment by jramosg** [comment]
@PEZ shall I merge this? Looks good to me now

**Comment by PEZ** [comment]
I lost track a bit. But as long as #2655 is safe, I am good with merging.
